src/probnum/kernels/_kernel.py

```diff
@@ -0,0 +1,302 @@
+    ) -> Tuple[np.ndarray, np.ndarray, bool]:
+        """Transform inputs to the kernel matrix.
+
+        Transforms inputs into :class:`numpy.ndarray` and standardizes their shape.
+
+        Parameters
+        ----------
+        x0 :
+            First input to the covariance function.
+        x1 :
+            Second input to the covariance function.
+
+        Returns
+        -------
+        x0 :
+            First input to the covariance function.
+        x1 :
+            Second input to the covariance function.
+        equal_inputs :
+            Are the two inputs the same?
+
+        Raises
+        -------
+        ValueError :
+            If input shapes of x0 and x1 do not match the kernel input dimension or
+            each other.
+        """
+        # pylint: disable="too-many-boolean-expressions"
+        # Transform into array and add second argument
+        x0 = np.asarray(x0)
+        equal_inputs = False
+        if x1 is None:
+            x1 = x0
+            equal_inputs = True
+        else:
+            x1 = np.asarray(x1)
+
+        # Error message
+        err_msg = (
+            f"Argument shapes x0.shape={x0.shape} and x1.shape="
+            f"{x1.shape} do not match kernel input dimension "
+            f"{self.input_dim}. Try passing either two vectors or two matrices with the "
+            f"second dimension equalling the kernel input dimension."
+        )
+
+        # Check and promote shapes
+        if x1 is None:
+
+            if (
+                (x0.ndim == 0 and self.input_dim > 1)  # Scalar input
+                or (x0.ndim == 1 and x0.shape[0] != self.input_dim)  # Vector input
+                or (x0.ndim >= 2 and x0.shape[1] != self.input_dim)  # Matrix input
+            ):
+                raise ValueError(err_msg)
+            equal_inputs = True
+        else:
+            # Promote unequal shapes
+            if x0.ndim < 2 and x1.ndim == 2:
+                x0 = np.atleast_2d(x0)
+            if x1.ndim < 2 and x0.ndim == 2:
+                x1 = np.atleast_2d(x1)
+            if x0.ndim != x1.ndim:  # Shape mismatch
+                raise ValueError(err_msg)
+
+            # Check shapes
+            if (
+                (x0.ndim == 0 and self.input_dim > 1)  # Scalar input
+                or (
+                    x0.ndim == 1  # Vector input
+                    and not (x0.shape[0] == x1.shape[0] == self.input_dim)
+                )
+                or (
+                    x0.ndim == 2  # Matrix input
+                    and not (x0.shape[1] == x1.shape[1] == self.input_dim)
+                )
+            ):
+                raise ValueError(err_msg)
+
+        return x0, x1, equal_inputs
+
+    def _transform_kernelmatrix(
+        self,
+        kerneval: Union[np.ndarray, ScalarArgType],
+        x0_shape: ShapeArgType,
+        x1_shape: ShapeArgType,
+    ) -> Union[np.float_, np.ndarray]:
+        """Transform the kernel matrix based on the given arguments.
+
```

Comment:
I am not sure I understand what this function does and why it is necessary, sorry to say this so bluntly...

Reply:
I changed it somewhat and tried to make the docstring more clear.